cpp/src/request_tag.cpp

```diff
@@ -148,26 +151,43 @@ void RequestTag::populateDelayedSubmission()
 
   request();
 
-  if (_enablePythonFuture)
-    ucxx_trace_req_f(_ownerString.c_str(),
-                     _request,
-                     _operationName.c_str(),
-                     "tag 0x%lx, tagMask: 0x%lx, buffer %p, size %lu, future %p, future handle %p, "
-                     "populateDelayedSubmission",
-                     _delayedSubmission->_data.getTag()._tag,
-                     _delayedSubmission->_data.getTag()._tagMask,
-                     _delayedSubmission->_buffer,
-                     _delayedSubmission->_length,
-                     _future.get(),
-                     _future->getHandle());
-  else
-    ucxx_trace_req_f(_ownerString.c_str(),
-                     _request,
-                     _operationName.c_str(),
-                     "tag 0x%lx, buffer %p, size %lu, populateDelayedSubmission",
-                     _delayedSubmission->_data.getTag()._tag,
-                     _delayedSubmission->_buffer,
-                     _delayedSubmission->_length);
+  auto log = [this](const void* buffer, const size_t length, const Tag tag, const TagMask tagMask) {
+    if (_enablePythonFuture)
+      ucxx_trace_req_f(
+        _ownerString.c_str(),
+        _request,
+        _operationName.c_str(),
+        "buffer: %p, size: %lu, tag 0x%lx, tagMask: 0x%lx, future %p, future handle %p, "
+        "populateDelayedSubmission",
+        buffer,
+        length,
+        tag,
+        tagMask,
+        _future.get(),
+        _future->getHandle());
+    else
+      ucxx_trace_req_f(
+        _ownerString.c_str(),
+        _request,
+        _operationName.c_str(),
+        "buffer: %p, size: %lu, tag 0x%lx, tagMask: 0x%lx, populateDelayedSubmission",
+        buffer,
+        length,
+        tag,
+        tagMask);
+  };
+
+  try {
+    auto tagSend = _delayedSubmission->_data.getTagSend();
+    log(tagSend._buffer, tagSend._length, tagSend._tag, TagMaskFull);
+  } catch (const std::bad_variant_access& e) {
+    try {
+      auto tagReceive = _delayedSubmission->_data.getTagReceive();
+      log(tagReceive._buffer, tagReceive._length, tagReceive._tag, tagReceive._tagMask);
+    } catch (const std::bad_variant_access& e) {
+      ucxx_error("Impossible to get transfer data.");
+    }
+  }
```

Comment:
This pattern could then also be wrapped up in the same `std::visit` pattern.

Reply:
Also done via https://github.com/rapidsai/ucxx/pull/121/commits/e80759b720ac66c7cf71aaeb78143cd6bcadeab9 .